### PR TITLE
Zstd support

### PR DESCRIPTION
**Pull request by SkBlaz**
Adding support for `.zst`, it appears this works at least as fast also.

### CODE REVIEW

File: src/main.rs [hunk 1]
```diff
@@ -15,36 +15,70 @@ use std::io::BufWriter;
...
+        .and_then(|ext| ext.to_str())
+        .expect("Failed to get the file extension.");
+
+    match input_format {
+        "gz" => {
+            let gz_decoder = MultiGzDecoder::new(input);
+            let reader = io::BufReader::new(gz_decoder);
+            Box::new(reader)
+        }
+        "zst" => {
```
Comment (ggaspersic) [inline src/main.rs:67]: We should add tests when adding a new input/output format

Comment (SkBlaz) [reply src/main.rs:67]: Added, good point

Comment (ggaspersic) [reply src/main.rs:67]: Github knows you didn't add tests, now that you did the build passes :)

**Review by ggaspersic — APPROVED** [review]

File: src/main.rs (hunk 1 above)
Comment (SkBlaz) [reply src/main.rs:67]: Unfortunately no, build issue was due most likely due to stack/runner conflict, put that on todos

File: src/main.rs [hunk 2]
```diff
@@ -300,10 +309,118 @@ fn main2() -> Result<(), Box<dyn Error>> {
...
+                quantize_weights,
+            )
+            .unwrap()
         }
     }
 
     Ok(())
 }
+
+#[cfg(test)]
```
Comment (bbenshalom) [inline src/main.rs:326]: Not sure I like these very specific tests in the main module... Can we extract the compression logic into another module?

File: src/block_ffm.rs [hunk 3]
```diff
@@ -1937,7 +1937,7 @@ mod tests {
             contra_field_index: mi.ffm_k,
         }]);
         assert_eq!(spredict2(&mut bg, &fb, &mut pb), 0.5);
-        assert_eq!(slearn2(&mut bg, &fb, &mut pb, true), 0.5);
+        assert_eq!(slearn2(&mut bg, &fb, &mut pb, true), 0.62245935);
```
Comment (bbenshalom) [inline src/block_ffm.rs:1940]: why would the expected result in this test change?

**Review by bbenshalom — CHANGES_REQUESTED** [review]

File: src/block_ffm.rs (hunk 3 above)
Comment (SkBlaz) [reply src/block_ffm.rs:1940]: Actually, current `main` fails with the same issue, so imagining something changed at actions/CI level.

Comment (SkBlaz) [reply src/block_ffm.rs:1940]: related to stack increases it appears.

File: src/main.rs (hunk 2 above)
Comment (SkBlaz) [reply src/main.rs]: Separated to `buffer_handler.rs` + applied `rustfmt` to align the repo

**Review by bbenshalom — APPROVED** [review]

**Pull request merged**